javascript/ql/src/semmle/javascript/Closure.qll

```diff
@@ -37,39 +37,162 @@ class GoogFunctionCallStmt extends ExprStmt {
+  GoogModuleDeclaration() {
+    getFunctionName() = "module" or
+    getFunctionName() = "declareModuleId"
+  }
+
+  /** Gets the identifier of the namespace imported by this call. */
+  override string getNamespaceId() { result = getArgument(0).(ConstantString).getStringValue() }
+}
+
+/**
+ * A module using the Closure module system, declared using `goog.module()` or `goog.declareModuleId()`.
+ */
+class ClosureModule extends Module {
   ClosureModule() {
+    getAChildStmt() instanceof GoogModuleDeclaration
+  }
+
+  /**
+   * Gets the call to `goog.module()` or `goog.declareModuleId` in this module.
+   */
+  GoogModuleDeclaration getModuleDeclaration() {
+    result = getAChildStmt()
+  }
+
+  /**
+   * Gets the namespace of this module.
+   */
+  string getNamespaceId() { result = getModuleDeclaration().getNamespaceId() }
+
+  override Module getAnImportedModule() {
+    exists (GoogRequireImport imprt | 
+      imprt.getEnclosingModule() = this and
+      result.(ClosureModule).getNamespaceId() = imprt.getNamespaceId()
+    )
+  }
+
+  /**
+   * Gets the top-level `exports` variable in this module, if this module is defined by
+   * a `good.module` call.
+   *
+   * This variable denotes the object exported from this module.
+   *
+   * Has no result for ES6 modules using `goog.declareModuleId`.
+   */
+  Variable getExportsVariable() {
+    getModuleDeclaration().getFunctionName() = "module" and
+    result.getScope() = this.getScope() and
+    result.getName() = "exports"
+  }
+
+  override predicate exports(string name, ASTNode export) {
+    // exports.foo = bar
+    export.(AssignExpr).getLhs().(PropAccess).accesses(getExportsVariable().getAnAccess(), name)
+    or
+    // exports = { foo: bar }
+    exists (VarDef def |
+      def.getTarget() = getExportsVariable().getAReference() and
+      def.getSource().(ObjectExpr).getPropertyByName(name) = export
+    )
+  }
+}
+
+/**
+ * A global Closure script, that is, a toplevel that is executed in the global scope and
+ * contains a toplevel call to `goog.provide` or `goog.require`.
+ */
+class ClosureScript extends TopLevel {
+  ClosureScript() {
+    not this instanceof ClosureModule and
     getAChildStmt() instanceof GoogProvide or
-    getAChildStmt() instanceof GoogRequire
+    getAChildStmt().(ExprStmt).getExpr() instanceof GoogRequire
   }
 
   /** Gets the identifier of a namespace required by this module. */
-  string getARequiredNamespace() { result = getAChildStmt().(GoogRequire).getNamespaceId() }
+  string getARequiredNamespace() { result = getAChildStmt().(ExprStmt).getExpr().(GoogRequire).getNamespaceId() }
 
   /** Gets the identifer of a namespace provided by this module. */
   string getAProvidedNamespace() { result = getAChildStmt().(GoogProvide).getNamespaceId() }
 }
+
+/**
+ * Holds if `name` is a closure namespace, including proper namespace prefixes.
+ */
+pragma[noinline]
+predicate isClosureLibraryNamespacePath(string name) {
```

Comment:
We generally avoid putting predicates directly into the global namespace. Perhaps this argues for introducing a `Closure` module? We could put the above classes in there as well, and keep the old names as deprecated aliases.

Reply:
👍 

It wasn't imported from `javascript.qll`, but that was admittedly a silly way to encapsulate it when we can just use a module.